server/src/main/scala/ch.epfl.scala.index.server/DataRepository.scala

```diff
@@ -1,667 +1,615 @@
         .toList
         .sortBy(_._1)
         .map { case (v, c) => (v.toString, c) }
     }
-
-    aggregations(field, params).map(sortedByVersion)
   }
 
   private def aggregations(
       field: String,
-      params: Option[SearchParams]
+      query: QueryDefinition
   ): Future[Map[String, Long]] = {
-    import scala.collection.JavaConverters._
     val aggregationName = s"${field}_count"
 
-    val q = params.map(getQuery).getOrElse(matchAllQuery)
+    val aggregation = termsAggregation(aggregationName).field(field).size(50)
 
-    esClient
-      .execute {
-        search(indexName / projectsCollection)
-          .query(q)
-          .aggregations(
-            termsAggregation(aggregationName)
-              .field(field)
-              .size(50)
-          )
-      }
-      .map(resp => {
-        try {
-          val agg = resp.aggregations.stringTermsResult(aggregationName)
-          agg.getBuckets.asScala.toList.collect {
-            case b: StringTerms.Bucket => b.getKeyAsString -> b.getDocCount
-          }.toMap
-        } catch {
-          case e: Exception => {
-            log.error("failed aggregations", e)
-            Map()
-          }
+    val request = search(indexName / projectsCollection)
+      .query(query)
+      .aggregations(aggregation)
+
+    for (response <- esClient.execute(request)) yield {
+      response.aggregations
+        .stringTermsResult(aggregationName)
+        .getBuckets
+        .asScala
+        .map { bucket =>
+          bucket.getKeyAsString -> bucket.getDocCount
         }
-      })
+        .toMap
+    }
   }
+}
 
-  private lazy val cachedTopics: Set[String] = {
-    import scala.collection.JavaConverters._
+object DataRepository {
+  private def gitHubStarScoring(query: QueryDefinition): QueryDefinition = {
+    val scorer = fieldFactorScore("github.stars")
+      .missing(0)
+      .modifier(Modifier.LN2P)
+    functionScoreQuery(query)
+      .scorers(scorer)
+      .boostMode(CombineFunction.MULTIPLY)
+  }
 
-    val aggregationName = "topic_count"
+  private def filteredSearchQuery(params: SearchParams): QueryDefinition = {
+    must(
+      notDeprecatedQuery,
+      repositoriesQuery(params.userRepos.toSeq),
+      optionalQuery(params.cli, cliQuery),
+      topicsQuery(params.topics),
+      targetsQuery(
+        params.targetTypes,
+        params.scalaVersions,
+        params.scalaJsVersions,
+        params.scalaNativeVersions,
+        params.sbtVersions
+      ),
+      optionalQuery(params.targetFiltering)(targetQuery),
+      optionalQuery(params.contributingSearch, contributingQuery),
+      searchQuery(params.queryString, params.contributingSearch)
+    )
+  }
 
-    Await.result(
-      esClient
-        .execute {
-          search(indexName / projectsCollection)
-            .aggregations(
-              termsAggregation(aggregationName).field("github.topics")
-            )
+  private def sortQuery(sorting: Option[String]): SortDefinition =
```

Comment:
Should we use an enumeration rather than a `String`?

Reply:
It would be nice because easier to read and less errorprone. Yet the sorting argument come from an http parameter and it could contain any string. There must be a mapping from string to something somewhere. This code was already there and I dont think it is a priority to change it now.

Reply:
Indeed, sorry I didn’t realize it was already here before.

Reply:
Any comment is appreciated